File: sdk/reproit-backend-rs/examples/hermetic_fixture.rs
//! Money-test fixture, Rust flavor: an axum app whose GET /quote operation
//! 500s because an upstream pricing service returns `{"prices": null}` and
//! the handler indexes into it.
//!
//! MODE=capture: boots the upstream, runs the failing operation through the
//! instrument boundaries with a standalone recorder, and writes a version-2
//! `reproit-backend-capture` (exchanges + envelope) to CAPTURE_OUT.
//! Default (server) mode: binds ONLY the app on $PORT; with REPROIT_REPLAY
//! set the SDK serves the recorded exchanges, so no upstream and no
//! database exist. FIXED=1 applies the fix.

use axum::routing::get;
use axum::Router;
use reproit_backend::instrument::db::{DbError, DbOutcome};
use reproit_backend::instrument::{self, db, http};
use reproit_backend::{BackendTrace, Recorder, TraceContext};
use serde_json::{json, Value};

async fn pg_lookup(symbol: &str) -> Result<DbOutcome, DbError> {
    let symbol = symbol.to_string();
    db::run(
        "SELECT id, symbol FROM issuers WHERE symbol = $1",
        &[json!(symbol)],
        || async {
            if std::env::var("MODE").as_deref() != Ok("capture") {
                return Err(DbError {
                    message: "live database reached during hermetic replay".into(),
                    code: None,
                });
            }
            Ok(DbOutcome {
                command: Some("SELECT".into()),
                row_count: 1,
                rows: vec![json!({"id": 7, "symbol": symbol})],
            })
        },
    )
    .await
}

/// The planted operation: (status, output) exactly like the Node fixture.
async fn quote(client: &reqwest::Client, upstream: &str, symbol: &str) -> (u16, Value) {
    if pg_lookup(symbol).await.is_err() {
        return (500, json!({"error": "internal"}));
    }
    let request = match client.get(format!("{upstream}/prices?tier=gold")).build() {
        Ok(request) => request,
        Err(_) => return (500, json!({"error": "internal"})),
    };
    let Ok(response) = http::send(client, request).await else {
        return (500, json!({"error": "internal"}));
    };
    let Ok(body) = response.json::<Value>() else {
        return (500, json!({"error": "internal"}));
    };
    let prices = &body["prices"];
    if std::env::var("FIXED").as_deref() == Ok("1") && !prices.is_array() {
        return (200, json!({"first": null, "note": "no prices available"}));
    }
    match prices.as_array().and_then(|prices| prices.first()) {
        Some(first) => (200, json!({"first": first})),
        None => (500, json!({"error": "internal"})),
    }
}

async fn capture_mode() {
    let upstream_app = Router::new().route(
        "/prices",
        get(|| async { axum::Json(json!({"prices": null})) }),
    );
    let upstream_listener = tokio::net::TcpListener::bind("127.0.0.1:19971")
        .await
        .expect("bind upstream");
    tokio::spawn(async move {
        axum::serve(upstream_listener, upstream_app)
            .await
            .expect("serve upstream");
    });

    let context = TraceContext {
        trace_id: "cap-money-rs-1".into(),
        actor: None,
        action_index: 0,
        build: Some("money-fixture".into()),
        config_contract: None,
        capture_envelope: true,
    };
    let trace = BackendTrace::begin(
        context,
        "GET /quote",
        None,
        None,
        None,
        json!({"query": {"symbol": "ACME"}}),
        Vec::new(),
    )
    .expect("begin trace");
    let recorder = Recorder::standalone(trace);
    let client = reqwest::Client::new();
    let (status, output) = instrument::scope(
        recorder.clone(),
        quote(&client, "http://127.0.0.1:19971", "ACME"),
    )
    .await;
    let mut trace = recorder.into_trace().expect("trace back");
    trace
        .finish(output, status, status < 500, true)
        .expect("finish trace");
    let payload = json!({
        "format": "reproit-backend-capture",
        "version": 2,
        "operation": "GET /quote",
        "oracle": "backend-server-error",
        "envelope": {
            "observedAtMs": std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|elapsed| elapsed.as_millis() as u64)
                .unwrap_or(0),
            "runtime": "rust",
            "os": std::env::consts::OS,
            "arch": std::env::consts::ARCH,
            "replaySeed": "c0ffee00c0ffee00",
        },
        "events": trace.events(),
    });
    let out = std::env::var("CAPTURE_OUT").expect("CAPTURE_OUT");
    std::fs::write(&out, serde_json::to_vec(&payload).expect("payload")).expect("write payload");
    println!("capture fixture status {status}");
}

async fn server_mode() {
    instrument::init();
    let client = reqwest::Client::new();
    let app = Router::new().route(
        "/quote",
        get({
            let client = client.clone();
            move |query: axum::extract::RawQuery| async move {
                let symbol = query
                    .0
                    .as_deref()
                    .and_then(|raw| {
                        raw.split('&')
                            .find_map(|pair| pair.strip_prefix("symbol=").map(str::to_string))
                    })
                    .unwrap_or_default();
                let (status, output) = quote(&client, "http://127.0.0.1:19971", &symbol).await;
                (
                    axum::http::StatusCode::from_u16(status)
                        .unwrap_or(axum::http::StatusCode::INTERNAL_SERVER_ERROR),
                    axum::Json(output),
                )
            }
        }),
    );
    let port = std::env::var("PORT")
        .ok()
        .and_then(|port| port.parse::<u16>().ok())
        .unwrap_or(19970);
    let listener = tokio::net::TcpListener::bind(("127.0.0.1", port))
        .await
        .expect("bind app");
    println!("serving on {port}");
    axum::serve(listener, app).await.expect("serve app");
}

#[tokio::main]
async fn main() {
    if std::env::var("MODE").as_deref() == Ok("capture") {
        capture_mode().await;
    } else {
        server_mode().await;
    }
}

File: sdk/reproit-backend-java/src/test/java/dev/reproit/backend/HermeticFixture.java
// Money-test fixture: a real Jetty container running ReproitFilter whose
// /quote operation 500s because an upstream pricing service returns
// {"prices": null} and the handler indexes into it.
//
// MODE=capture: boots the upstream and the app, fires the failing request,
// and writes a version-2 reproit-backend-capture (exchanges + envelope) to
// CAPTURE_OUT. Default (server) mode boots ONLY the app on $PORT; with
// REPROIT_REPLAY set the SDK serves the recorded exchanges, so no upstream
// and no database exist. FIXED=1 applies the fix.
package dev.reproit.backend;

import com.sun.net.httpserver.HttpServer;
import jakarta.servlet.DispatcherType;
import jakarta.servlet.http.HttpServlet;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.servlet.FilterHolder;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;

public final class HermeticFixture {
    private static final int UPSTREAM_PORT = 19981;
    private static final String UPSTREAM =
        "http://127.0.0.1:" + UPSTREAM_PORT + "/prices?tier=gold";
    private static final boolean FIXED = "1".equals(System.getenv("FIXED"));

    private HermeticFixture() {}

    /** The /quote handler: one db read, one upstream call, then the defect. */
    public static final class QuoteServlet extends HttpServlet {
        @Override
        protected void doGet(HttpServletRequest request, HttpServletResponse response)
                throws IOException {
            response.setContentType("application/json");
            try {
                Instrument.Db.run(
                    "SELECT id, symbol FROM issuers WHERE symbol = $1",
                    List.of(String.valueOf(request.getParameter("symbol"))),
                    () -> {
                        if (!"capture".equals(System.getenv("MODE"))) {
                            throw new IllegalStateException(
                                "live database reached during hermetic replay");
                        }
                        return new Instrument.Db.Outcome(
                            "SELECT", 1, List.of(Map.of("id", 7L, "symbol", "ACME")));
                    });
                Instrument.Http.ExchangeResponse upstream = Instrument.Http.send(
                    HttpClient.newHttpClient(),
                    HttpRequest.newBuilder(URI.create(UPSTREAM)).GET().build());
                Object prices = ((Map<?, ?>) upstream.json()).get("prices");
                if (FIXED && !(prices instanceof List)) {
                    response.getWriter().write("{\"first\":null,\"note\":\"no prices\"}");
                    return;
                }
                Object first = ((List<?>) prices).get(0);
                response.getWriter().write("{\"first\":" + first + "}");
            } catch (RuntimeException | InterruptedException | IOException failure) {
                response.setStatus(500);
                response.getWriter().write("{\"error\":\"internal\"}");
            }
        }
    }

    public static void main(String[] args) throws Exception {
        Instrument.init();
        boolean capturing = "capture".equals(System.getenv("MODE"));
        HttpServer upstream = null;
        if (capturing) {
            upstream = HttpServer.create(new InetSocketAddress("127.0.0.1", UPSTREAM_PORT), 0);
            upstream.createContext("/prices", exchange -> {
                byte[] reply = "{\"prices\":null}".getBytes(StandardCharsets.UTF_8);
                exchange.getResponseHeaders().set("Content-Type", "application/json");
                exchange.sendResponseHeaders(200, reply.length);
                exchange.getResponseBody().write(reply);
                exchange.close();
            });
            upstream.start();
        }

        List<BackendTrace> recorded = new ArrayList<>();
        Server server = new Server();
        ServerConnector connector = new ServerConnector(server);
        connector.setHost("127.0.0.1");
        connector.setPort(capturing ? 19980 : Integer.parseInt(
            System.getenv().getOrDefault("PORT", "19980")));
        server.addConnector(connector);
        ServletContextHandler handler = new ServletContextHandler();
        handler.setContextPath("/");
        // The filter runs without a Capture: the fixture writes the capture
        // file itself so the money test needs no cloud endpoint.
        handler.addFilter(
            new FilterHolder(new ReproitFilter(new FileSink(recorded))), "/*",
            EnumSet.of(DispatcherType.REQUEST));
        handler.addServlet(new ServletHolder(new QuoteServlet()), "/quote");
        server.setHandler(handler);
        server.start();

        if (!capturing) {
            server.join();
            return;
        }
        String base = "http://127.0.0.1:" + connector.getLocalPort();
        HttpResponse<String> failing = HttpClient.newHttpClient().send(
            HttpRequest.newBuilder(URI.create(base + "/quote?symbol=ACME")).GET().build(),
            HttpResponse.BodyHandlers.ofString());
        System.out.println("capture fixture status " + failing.statusCode());
        writeCapture(recorded.get(recorded.size() - 1));
        server.stop();
        upstream.stop(0);
        System.exit(0);
    }

    /**
     * Supplies the capture-mode context and keeps the finished trace in
     * memory, so the money test needs no cloud endpoint.
     */
    private static final class FileSink implements TraceSink {
        private final List<BackendTrace> traces;

        FileSink(List<BackendTrace> traces) {
            this.traces = traces;
        }

        @Override
        public TraceContext context() {
            return new TraceContext("cap-java-1", null, 0, "money-fixture", null, true);
        }

        @Override
        public void record(BackendTrace trace) {
            traces.add(trace);
        }
    }

    static void writeCapture(BackendTrace trace) throws IOException {
        Map<String, Object> envelope = Capture.determinismEnvelope(null);
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("format", Capture.CAPTURE_FORMAT);
        payload.put("version", 2);
        payload.put("operation", trace.events().get(0).get("operation"));
        payload.put("oracle", Capture.SERVER_ERROR_ORACLE);
        payload.put("envelope", envelope);
        payload.put("events", trace.events());
        Files.writeString(
            Path.of(System.getenv("CAPTURE_OUT")),
            Json.canonicalJson(payload),
            StandardCharsets.UTF_8);
    }
}
